src/witan/app/forecast.clj

```diff
@@ -10,97 +10,148 @@
+
+
+(defn create-forecast-name
+  [{:keys [name owner]}]
+  (hayt/insert :forecast_names (hayt/values :name name
+                                            :owner owner)))
+
+(defn create-forecast-version
+  [{:keys [name description owner forecast-id version in_progress id version-id]}]
+  (let [creation-time (tf/unparse (tf/formatters :date-time) (t/now))]
+    (hayt/insert :forecasts (hayt/values
+                                     :forecast_id forecast-id
+                                     :name  name
+                                     :description description
+                                     :created creation-time
+                                     :owner owner
+                                     :version_id version-id
+                                     :version version
+                                     :in_progress in_progress))))
+(defn create-first-version
+  [{:keys [forecast-id version-id name description owner]}]
+  (create-forecast-version {:name name
+                            :description description
+                            :forecast-id forecast-id
+                            :version-id version-id
+                            :version 0
+                            :owner owner
+                            :in_progress false}))
 
 (defn add-forecast!
   [{:keys [name owner] :as forecast}]
   (let [existing-forecasts (c/exec (find-forecast-by-name-and-owner name owner))
-        id (uuid/random)]
+        id (uuid/random)
+        version-id (uuid/random)
+        new-forecast (assoc forecast :forecast-id id :version-id version-id)]
     (when (empty? existing-forecasts)
-      (c/exec (create-forecast (assoc forecast :id id)))
+      (c/exec (create-new-forecast new-forecast))
+      (c/exec (create-first-version new-forecast))
+      (c/exec (create-forecast-name new-forecast))
       (first (c/exec (find-forecast-by-id id))))))
 
 (defn update-forecast!
-  [{:keys [series-id owner inputs]}]
-  (if-let [latest-forecast (retrieve-forecast-most-recent-of-series series-id)]
+  [{:keys [id owner]}]
+  (if-let [latest-forecast (retrieve-forecast-most-recent-of-series id)]
     (let [new-version (inc (:version latest-forecast))
-          new-id (uuid/random)
+          new-version-id (uuid/random)
           new-forecast (assoc latest-forecast
                               :version new-version
+                              :version-id new-version-id
                               :owner owner
-                              :in_progress true
-                              :id new-id)]
-      (c/exec (create-forecast new-forecast))
-      (c/exec (update-forecast-descendant-id (:id latest-forecast) new-id))
-      (first (c/exec (find-forecast-by-id new-id))))))
+                              :in-progress true
+                              :forecast-id id)]
+      (c/exec (create-forecast-version new-forecast))
+      (c/exec (update-forecast-current-version-id id new-version-id new-version)))))
```

Comment:
I'd go so far as to say I *definitely* need it. The flow will be, you press "Update" on the forecast, the POST is sent, a 200 is returned and then what? How will I know the ID of the new forecast?